https://github.com/canonical/cloud-init/commit/a925b5a0ca4aa3e63b084c0f6664fe815c2c9db0

=== Begin SRU Template ===
[Impact]
The APT GPG command would not work in environments that have no tty.

[Test Case]

There is an upstream test for this bug:
https://github.com/canonical/cloud-init/blob/master/tests/integration_tests/bugs/test_lp1813396.py

=== Xenial ===
============================= test session starts ==============================
platform linux -- Python 3.5.9, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: /home/james/cloud-init, inifile: tox.ini
plugins: cov-2.9.0
collected 1 item

tests/integration_tests/bugs/test_lp1813396.py Detected image: image_id=xenial os=ubuntu release=xenial
Detected image: image_id=xenial os=ubuntu release=xenial
Settings:
CLOUD_INIT_SOURCE=PROPOSED
COLLECT_LOGS=NEVER
EXISTING_INSTANCE_ID=None
GCE_PROJECT=ubuntu-server-277015
GCE_REGION=us-central1
GCE_ZONE=a
INSTANCE_TYPE=None
KEEP_IMAGE=False
KEEP_INSTANCE=0
LOCAL_LOG_PATH=/tmp/cloud_init_test_logs
OCI_COMPARTMENT_ID=ocid1.compartment.oc1..aaaaaaaanz4b63fdemmuag77dg2pi22xfyhrpq46hcgdd3dozkvqfzwwjwxa
OS_IMAGE=xenial
PLATFORM=lxd_container
RUN_UNSTABLE=True
Setting up environment for lxd_container
Launching instance with launch_kwargs:
image_id=ubuntu:de03ec8f36f75815b6b6c6dcae2dc0c7b09079b035a27c46ff854b0977f9139b
user_data=None
wait=False
['lxc', 'init', 'ubuntu:de03ec8f36f75815b6b6c6dcae2dc0c7b09079b035a27c46ff854b0977f9139b']
Launched instance: LXDInstance(name=faithful-gazelle)
Installing proposed image
Installed cloud-init version: 20.4-0ubuntu1~16.04.1
Created new image: local:faithful-gazelle-snapshot
Done with environment setup
Launching instance with launch_kwargs:
image_id=local:faithful-gazelle-snapshot
user_data=#cloud-config
apt:
  sources:
    cloudinit:
      source: 'deb [arch=amd64] http://ppa.launchpad.net/cloud-init-dev/daily/ubuntu focal main'
      keyserver: keyserver.ubuntu.com
      keyid: E4D304DF

wait=False
['lxc', 'init', 'local:faithful-gazelle-snapshot', '--config', "user.user-data=#cloud-config\napt:\n  sources:\n    cloudinit:\n      source: 'deb [arch=amd64] http://ppa.launchpad.net/cloud-init-dev/daily/ubuntu focal main'\n      keyserver: keyserver.ubuntu.com\n      keyid: E4D304DF\n"]
Launched instance: LXDInstance(name=wise-sheep)
.Deleting snapshot image created for this testrun: local:faithful-gazelle-snapshot


========================= 1 passed in 94.61s (0:01:34) =========================

=== Bionic ===
============================= test session starts ==============================
platform linux -- Python 3.5.9, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: /home/james/cloud-init, inifile: tox.ini
plugins: cov-2.9.0
collected 1 item

tests/integration_tests/bugs/test_lp1813396.py Detected image: image_id=bionic os=ubuntu release=bionic
Detected image: image_id=bionic os=ubuntu release=bionic
Settings:
CLOUD_INIT_SOURCE=PROPOSED
COLLECT_LOGS=NEVER
EXISTING_INSTANCE_ID=None
GCE_PROJECT=ubuntu-server-277015
GCE_REGION=us-central1
GCE_ZONE=a
INSTANCE_TYPE=None
KEEP_IMAGE=False
KEEP_INSTANCE=0
LOCAL_LOG_PATH=/tmp/cloud_init_test_logs
OCI_COMPARTMENT_ID=ocid1.compartment.oc1..aaaaaaaanz4b63fdemmuag77dg2pi22xfyhrpq46hcgdd3dozkvqfzwwjwxa
OS_IMAGE=bionic
PLATFORM=lxd_container
RUN_UNSTABLE=True
Setting up environment for lxd_container
Launching instance with launch_kwargs:
image_id=ubuntu:95c0e536d361eb5ac953ad343e0342c2f615e4aea714ca8a64126a228b809cae
wait=False
user_data=None
['lxc', 'init', 'ubuntu:95c0e536d361eb5ac953ad343e0342c2f615e4aea714ca8a64126a228b809cae']
Launched instance: LXDInstance(name=premium-quagga)
Installing proposed image
Installed cloud-init version: 20.4-0ubuntu1~18.04.1
Created new image: local:premium-quagga-snapshot
Done with environment setup
Launching instance with launch_kwargs:
image_id=local:premium-quagga-snapshot
wait=False
user_data=#cloud-config
apt:
  sources:
    cloudinit:
      source: 'deb [arch=amd64] http://ppa.launchpad.net/cloud-init-dev/daily/ubuntu focal main'
      keyserver: keyserver.ubuntu.com
      keyid: E4D304DF

['lxc', 'init', 'local:premium-quagga-snapshot', '--config', "user.user-data=#cloud-config\napt:\n  sources:\n    cloudinit:\n      source: 'deb [arch=amd64] http://ppa.launchpad.net/cloud-init-dev/daily/ubuntu focal main'\n      keyserver: keyserver.ubuntu.com\n      keyid: E4D304DF\n"]
Launched instance: LXDInstance(name=capable-bass)
.Deleting snapshot image created for this testrun: local:premium-quagga-snapshot


========================= 1 passed in 96.92s (0:01:36) =========================

=== Focal ===
============================= test session starts ==============================
platform linux -- Python 3.5.9, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: /home/james/cloud-init, inifile: tox.ini
plugins: cov-2.9.0
collected 1 item

tests/integration_tests/bugs/test_lp1813396.py Detected image: image_id=focal os=ubuntu release=focal
Detected image: image_id=focal os=ubuntu release=focal
Settings:
CLOUD_INIT_SOURCE=PROPOSED
COLLECT_LOGS=NEVER
EXISTING_INSTANCE_ID=None
GCE_PROJECT=ubuntu-server-277015
GCE_REGION=us-central1
GCE_ZONE=a
INSTANCE_TYPE=None
KEEP_IMAGE=False
KEEP_INSTANCE=0
LOCAL_LOG_PATH=/tmp/cloud_init_test_logs
OCI_COMPARTMENT_ID=ocid1.compartment.oc1..aaaaaaaanz4b63fdemmuag77dg2pi22xfyhrpq46hcgdd3dozkvqfzwwjwxa
OS_IMAGE=focal
PLATFORM=lxd_container
RUN_UNSTABLE=True
Setting up environment for lxd_container
Launching instance with launch_kwargs:
wait=False
user_data=None
image_id=ubuntu:e0c3495ffd489748aa5151628fa56619e6143958f041223cb4970731ef939cb6
['lxc', 'init', 'ubuntu:e0c3495ffd489748aa5151628fa56619e6143958f041223cb4970731ef939cb6']
Launched instance: LXDInstance(name=oriented-seahorse)
Installing proposed image
Installed cloud-init version: 20.4-0ubuntu1~20.04.1
Created new image: local:oriented-seahorse-snapshot
Done with environment setup
Launching instance with launch_kwargs:
wait=False
user_data=#cloud-config
apt:
  sources:
    cloudinit:
      source: 'deb [arch=amd64] http://ppa.launchpad.net/cloud-init-dev/daily/ubuntu focal main'
      keyserver: keyserver.ubuntu.com
      keyid: E4D304DF

image_id=local:oriented-seahorse-snapshot
['lxc', 'init', 'local:oriented-seahorse-snapshot', '--config', "user.user-data=#cloud-config\napt:\n  sources:\n    cloudinit:\n      source: 'deb [arch=amd64] http://ppa.launchpad.net/cloud-init-dev/daily/ubuntu focal main'\n      keyserver: keyserver.ubuntu.com\n      keyid: E4D304DF\n"]
Launched instance: LXDInstance(name=sure-anemone)
.Deleting snapshot image created for this testrun: local:oriented-seahorse-snapshot


======================== 1 passed in 116.02s (0:01:56) =========================

=== Groovy ===
============================= test session starts ==============================
platform linux -- Python 3.5.9, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: /home/james/cloud-init, inifile: tox.ini
plugins: cov-2.9.0
collected 1 item

tests/integration_tests/bugs/test_lp1813396.py Detected image: image_id=groovy os=ubuntu release=groovy
Detected image: image_id=groovy os=ubuntu release=groovy
Settings:
CLOUD_INIT_SOURCE=PROPOSED
COLLECT_LOGS=NEVER
EXISTING_INSTANCE_ID=None
GCE_PROJECT=ubuntu-server-277015
GCE_REGION=us-central1
GCE_ZONE=a
INSTANCE_TYPE=None
KEEP_IMAGE=False
KEEP_INSTANCE=0
LOCAL_LOG_PATH=/tmp/cloud_init_test_logs
OCI_COMPARTMENT_ID=ocid1.compartment.oc1..aaaaaaaanz4b63fdemmuag77dg2pi22xfyhrpq46hcgdd3dozkvqfzwwjwxa
OS_IMAGE=groovy
PLATFORM=lxd_container
RUN_UNSTABLE=True
Setting up environment for lxd_container
Launching instance with launch_kwargs:
wait=False
user_data=None
image_id=ubuntu:bac1692e9ec7f5c5cee35d149ca789121bed238c1182df2134b5aefa3977f3e0
['lxc', 'init', 'ubuntu:bac1692e9ec7f5c5cee35d149ca789121bed238c1182df2134b5aefa3977f3e0']
Launched instance: LXDInstance(name=innocent-sailfish)
Installing proposed image
Installed cloud-init version: 20.4-0ubuntu1~20.10.1
Created new image: local:innocent-sailfish-snapshot
Done with environment setup
Launching instance with launch_kwargs:
wait=False
user_data=#cloud-config
apt:
  sources:
    cloudinit:
      source: 'deb [arch=amd64] http://ppa.launchpad.net/cloud-init-dev/daily/ubuntu focal main'
      keyserver: keyserver.ubuntu.com
      keyid: E4D304DF

image_id=local:innocent-sailfish-snapshot
['lxc', 'init', 'local:innocent-sailfish-snapshot', '--config', "user.user-data=#cloud-config\napt:\n  sources:\n    cloudinit:\n      source: 'deb [arch=amd64] http://ppa.launchpad.net/cloud-init-dev/daily/ubuntu focal main'\n      keyserver: keyserver.ubuntu.com\n      keyid: E4D304DF\n"]
Launched instance: LXDInstance(name=concrete-pony)
.Deleting snapshot image created for this testrun: local:innocent-sailfish-snapshot


======================== 1 passed in 112.28s (0:01:52) =========================

[Regression Potential]
The apt module will now work where it failed in the past.